plugins/sinks/shield/sink.go

```diff
@@ -0,0 +1,207 @@
+	Summary:     summary,
+	Tags:        []string{"grpc", "sink"},
+	SampleConfig: heredoc.Doc(`
+	# The hostname of the shield service
+	host: https://shield.com
+	# Additional headers send to shield, multiple headers value are separated by a comma
+	headers:
+	  X-Shield-Email: meteor@odpf.io
+      X-Other-Header: value1, value2
+	`),
+}
+
+type Sink struct {
+	plugins.BasePlugin
+	client Client
+	config Config
+	logger log.Logger
+}
+
+func New(c Client, logger log.Logger) plugins.Syncer {
+	s := &Sink{
+		logger: logger,
+		client: c,
+	}
+	s.BasePlugin = plugins.NewBasePlugin(info, &s.config)
+
+	return s
+}
+
+func (s *Sink) Init(ctx context.Context, config plugins.Config) error {
+	if err := s.BasePlugin.Init(ctx, config); err != nil {
+		return err
+	}
+
+	if err := s.client.Connect(ctx, s.config.Host); err != nil {
+		return fmt.Errorf("error connecting to host: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Sink) Sink(ctx context.Context, batch []models.Record) error {
+	for _, record := range batch {
+		metadata := record.Data()
+		s.logger.Info("sinking record to shield", "record", metadata.GetUrn())
+
+		shieldPayload, err := s.buildShieldPayload(metadata)
+		if err != nil {
+			return errors.Wrap(err, "failed to build shield payload")
+		}
+		if err = s.send(ctx, shieldPayload); err != nil {
+			return errors.Wrap(err, "error sending data")
+		}
+
+		s.logger.Info("successfully sinked record to shield", "record", metadata.Name)
+	}
+
+	return nil
+}
+
+func (s *Sink) Close() (err error) {
+	return
+	//TODO: Connection closes even when some records are unpiblished
+	//TODO: return s.client.Close()
+}
+
+func (s *Sink) send(ctx context.Context, record RequestPayload) error {
+	for hdrKey, hdrVal := range s.config.Headers {
+		hdrVals := strings.Split(hdrVal, ",")
+		for _, val := range hdrVals {
+			val = strings.TrimSpace(val)
+			md := metadata.New(map[string]string{hdrKey: val})
+			ctx = metadata.NewOutgoingContext(ctx, md)
+		}
+	}
+
+	metadataBytes, err := json.Marshal(record.Metadata)
+	if err != nil {
+		return err
+	}
+
+	var requestBody sh.UserRequestBody
+	requestBody.Name = record.Name
+	requestBody.Email = record.Email
+	requestBody.Metadata = new(structpb.Struct)
+
+	err = json.Unmarshal(metadataBytes, requestBody.Metadata)
+	if err != nil {
+		return err
+	}
+
+	_, err = s.client.UpdateUser(ctx, &sh.UpdateUserRequest{
+		Id:   requestBody.Email,
+		Body: &requestBody,
+	})
+	if err == nil {
+		return nil
+	}
+
+	if e, ok := status.FromError(err); ok {
+		err = fmt.Errorf("shield returns code %d: %v", e.Code(), e.Message())
+		switch e.Code() {
+		case codes.Unavailable:
+			return plugins.NewRetryError(err)
+		default:
+			return err
+		}
+	} else {
+		err = fmt.Errorf("not able to parse error returned %v", err)
+	}
+
+	return err
+}
+
+func (s *Sink) buildShieldPayload(resource *assetsv1beta2.Asset) (RequestPayload, error) {
+	data := resource.GetData()
+
+	mapdata, err := s.buildShieldData(data)
```

Comment:
we can rename `buildShieldData` to `getUserData` which might return error if `anypb.Any` is not a `User` struct.

And if it returns error that Asset is not a User, we can just skip processing it since `shield` only concerns about `User` data.